File: discrete-math/practice-03/artifacts/solve_pr3_tasks7_9.js
const fs = require("fs");
const path = require("path");

const outDir = __dirname;
const vars = ["A", "B", "C", "D"];

const not = (x) => (x ? 0 : 1);
const and = (...xs) => (xs.every(Boolean) ? 1 : 0);
const or = (...xs) => (xs.some(Boolean) ? 1 : 0);
const xor = (a, b) => ((a ? 1 : 0) ^ (b ? 1 : 0));
const impl = (a, b) => or(not(a), b);
const equiv = (a, b) => (a === b ? 1 : 0);
const nand = (a, b) => not(and(a, b));

function rowsFor(fn) {
  const rows = [];
  for (let i = 0; i < 16; i += 1) {
    const A = (i >> 3) & 1;
    const B = (i >> 2) & 1;
    const C = (i >> 1) & 1;
    const D = i & 1;
    rows.push({ i, A, B, C, D, ...fn(A, B, C, D) });
  }
  return rows;
}

const task7Rows = rowsFor((A, B, C, D) => {
  const implication = impl(A, B);
  const bxord = xor(B, D);
  const right = and(C, bxord);
  const f = equiv(implication, right);
  return { implication, bxord, right, f };
});

const task8Rows = rowsFor((A, B, C, D) => {
  const implication = impl(A, B);
  const sheffer = nand(C, B);
  const equivalence = equiv(implication, sheffer);
  const f = xor(equivalence, D);
  return { implication, sheffer, equivalence, f };
});

function bin(n) {
  return n.toString(2).padStart(4, "0");
}

function ones(pattern) {
  return [...pattern].filter((x) => x === "1").length;
}

function combine(a, b) {
  let diff = 0;
  let result = "";
  for (let i = 0; i < a.length; i += 1) {
    if (a[i] === b[i]) result += a[i];
    else if (a[i] !== "-" && b[i] !== "-") {
      diff += 1;
      result += "-";
    } else return null;
  }
  return diff === 1 ? result : null;
}

function covers(pattern, minterm) {
  const bits = bin(minterm);
  return [...pattern].every((bit, i) => bit === "-" || bit === bits[i]);
}

function quine(minterms) {
  let groups = new Map();
  const initial = minterms.map((m) => ({ pattern: bin(m), terms: [m] }));
  for (const item of initial) {
    const key = ones(item.pattern);
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(item);
  }

  const stages = [];
  const primeMap = new Map();
  let current = initial;

  while (current.length > 0) {
    const used = new Set();
    const nextMap = new Map();
    const byOnes = new Map();
    for (const item of current) {
      const key = ones(item.pattern.split("-").join(""));
      if (!byOnes.has(key)) byOnes.set(key, []);
      byOnes.get(key).push(item);
    }

    const stageRows = current.map((item) => ({ pattern: item.pattern, terms: item.terms }));
    const keys = [...byOnes.keys()].sort((a, b) => a - b);
    for (const key of keys) {
      const left = byOnes.get(key) || [];
      const right = byOnes.get(key + 1) || [];
      for (const a of left) {
        for (const b of right) {
          const pattern = combine(a.pattern, b.pattern);
          if (pattern) {
            used.add(a.pattern);
            used.add(b.pattern);
            const terms = [...new Set([...a.terms, ...b.terms])].sort((x, y) => x - y);
            if (!nextMap.has(pattern)) nextMap.set(pattern, { pattern, terms });
          }
        }
      }
    }

    for (const item of current) {
      if (!used.has(item.pattern)) primeMap.set(item.pattern, item);
    }
    stages.push(stageRows);
    current = [...nextMap.values()];
  }

  const primes = [...primeMap.values()]
    .map((p) => ({ pattern: p.pattern, terms: minterms.filter((m) => covers(p.pattern, m)) }))
    .sort((a, b) => a.pattern.localeCompare(b.pattern));

  const chart = Object.fromEntries(minterms.map((m) => [
    m,
    primes.filter((p) => covers(p.pattern, m)).map((p) => p.pattern),
  ]));

  const essential = [];
  for (const m of minterms) {
    if (chart[m].length === 1 && !essential.includes(chart[m][0])) essential.push(chart[m][0]);
  }

  const coversAll = (patterns) =>
    minterms.every((m) => patterns.some((pattern) => covers(pattern, m)));

  const remainingPrimes = primes.map((p) => p.pattern);
  const candidates = [];
  for (let mask = 1; mask < (1 << remainingPrimes.length); mask += 1) {
    const patterns = remainingPrimes.filter((_, i) => mask & (1 << i));
    if (coversAll(patterns)) candidates.push(patterns);
  }
  candidates.sort((a, b) =>
    literalCount(a) - literalCount(b)
    || a.length - b.length
    || a.join(",").localeCompare(b.join(",")));

  return { minterms, stages, primes, chart, essential, selected: candidates[0] };
}

function literalCount(patterns) {
  return patterns.reduce((total, pattern) => total + [...pattern].filter((x) => x !== "-").length, 0);
}

function dnfTerm(pattern) {
  const lits = [...pattern].flatMap((bit, i) => {
    if (bit === "-") return [];
    return bit === "1" ? [vars[i]] : [`¬${vars[i]}`];
  });
  return lits.length ? lits.join(" ∧ ") : "1";
}

function cnfClauseFromZero(pattern) {
  const lits = [...pattern].flatMap((bit, i) => {
    if (bit === "-") return [];
    return bit === "1" ? [`¬${vars[i]}`] : [vars[i]];
  });
  return lits.length ? `(${lits.join(" ∨ ")})` : "(0)";
}

function karnaugh(rows) {
  const abOrder = ["00", "01", "11", "10"];
  const cdOrder = ["00", "01", "11", "10"];
  return abOrder.map((ab) =>
    cdOrder.map((cd) => {
      const A = Number(ab[0]);
      const B = Number(ab[1]);
      const C = Number(cd[0]);
      const D = Number(cd[1]);
      return rows.find((r) => r.A === A && r.B === B && r.C === C && r.D === D).f;
    }));
}

const task8Minterms = task8Rows.filter((r) => r.f === 1).map((r) => r.i);
const task8Zeros = task8Rows.filter((r) => r.f === 0).map((r) => r.i);
const task8Dnf = quine(task8Minterms);
const task8Cnf = quine(task8Zeros);

const task9Minterms = [1, 2, 3, 5, 10, 14, 15];
const task9 = quine(task9Minterms);

const results = {
  task7: {
    variant: 14,
    expression: "(A → B) ~ (C ∧ (B ⊕ D))",
    rows: task7Rows,
    vector: task7Rows.map((r) => r.f),
  },
  task8: {
    variant: 1,
    expression: "((A → B) ~ (C | B)) ⊕ D",
    rows: task8Rows,
    vector: task8Rows.map((r) => r.f),
    minterms: task8Minterms,
    zeros: task8Zeros,
    karnaugh: karnaugh(task8Rows),
    minDnfPatterns: task8Dnf.selected,
    minDnf: task8Dnf.selected.map(dnfTerm).join(" ∨ "),
    minCnfPatterns: task8Cnf.selected,
    minCnf: task8Cnf.selected.map(cnfClauseFromZero).join(" ∧ "),
  },
  task9: {
    variant: 2,
    minterms: task9Minterms,
    stages: task9.stages,
    primeImplicants: task9.primes,
    chart: task9.chart,
    essential: task9.essential,
    selected: task9.selected,
    minDnf: task9.selected.map(dnfTerm).join(" ∨ "),
  },
};

const text = [
  "Практична робота 3: контрольні результати",
  "",
  `Задача 7: вектор f = [${results.task7.vector.join(", ")}]`,
  "",
  `Задача 8: мінтерми = (${task8Minterms.join(", ")}), нулі = (${task8Zeros.join(", ")})`,
  `  мінДНФ = ${results.task8.minDnf}`,
  `  мінКНФ = ${results.task8.minCnf}`,
  "",
  `Задача 9: F = (${task9Minterms.join(", ")})`,
  `  істотні імпліканти = ${task9.essential.join(", ")}`,
  `  мінДНФ = ${results.task9.minDnf}`,
  "",
].join("\n");

fs.writeFileSync(
  path.join(outDir, "pr3_tasks7_9_results.json"),
  JSON.stringify(results, null, 2),
  "utf8",
);
fs.writeFileSync(path.join(outDir, "pr3_tasks7_9_results.txt"), text, "utf8");

console.log(text);

File: discrete-math/practice-03/solve_pr3_tasks7_9.py
from __future__ import annotations

import json
import os
from pathlib import Path


OUT_DIR = Path(os.path.relpath(Path(__file__).parent, Path.cwd()))
VARS = ["A", "B", "C", "D"]


def logical_not(x: int) -> int:
    return 0 if x else 1


def logical_and(*xs: int) -> int:
    return 1 if all(xs) else 0


def logical_or(*xs: int) -> int:
    return 1 if any(xs) else 0


def xor(a: int, b: int) -> int:
    return int(bool(a)) ^ int(bool(b))


def impl(a: int, b: int) -> int:
    return logical_or(logical_not(a), b)


def equiv(a: int, b: int) -> int:
    return 1 if a == b else 0


def nand(a: int, b: int) -> int:
    return logical_not(logical_and(a, b))


def rows_for(fn):
    rows = []
    for i in range(16):
        a = (i >> 3) & 1
        b = (i >> 2) & 1
        c = (i >> 1) & 1
        d = i & 1
        rows.append({"i": i, "A": a, "B": b, "C": c, "D": d, **fn(a, b, c, d)})
    return rows


def bin4(n: int) -> str:
    return format(n, "04b")


def ones(pattern: str) -> int:
    return pattern.count("1")


def combine(a: str, b: str) -> str | None:
    diff = 0
    result = ""
    for left, right in zip(a, b):
        if left == right:
            result += left
        elif left != "-" and right != "-":
            diff += 1
            result += "-"
        else:
            return None
    return result if diff == 1 else None


def covers(pattern: str, minterm: int) -> bool:
    bits = bin4(minterm)
    return all(bit == "-" or bit == bits[index] for index, bit in enumerate(pattern))


def literal_count(patterns: list[str]) -> int:
    return sum(sum(1 for bit in pattern if bit != "-") for pattern in patterns)


def quine(minterms: list[int]) -> dict[str, object]:
    initial = [{"pattern": bin4(m), "terms": [m]} for m in minterms]
    stages = []
    prime_map: dict[str, dict[str, object]] = {}
    current = initial

    while current:
        used: set[str] = set()
        next_map: dict[str, dict[str, object]] = {}
        by_ones: dict[int, list[dict[str, object]]] = {}
        for item in current:
            key = ones(str(item["pattern"]).replace("-", ""))
            by_ones.setdefault(key, []).append(item)

        stage_rows = [{"pattern": item["pattern"], "terms": item["terms"]} for item in current]
        for key in sorted(by_ones):
            for left in by_ones.get(key, []):
                for right in by_ones.get(key + 1, []):
                    pattern = combine(str(left["pattern"]), str(right["pattern"]))
                    if pattern:
                        used.add(str(left["pattern"]))
                        used.add(str(right["pattern"]))
                        terms = sorted(set(left["terms"]) | set(right["terms"]))
                        next_map.setdefault(pattern, {"pattern": pattern, "terms": terms})

        for item in current:
            if item["pattern"] not in used:
                prime_map[str(item["pattern"])] = item
        stages.append(stage_rows)
        current = list(next_map.values())

    primes = sorted(
        [
            {"pattern": pattern, "terms": [m for m in minterms if covers(pattern, m)]}
            for pattern in prime_map
        ],
        key=lambda item: str(item["pattern"]),
    )

    chart = {
        m: [str(prime["pattern"]) for prime in primes if covers(str(prime["pattern"]), m)]
        for m in minterms
    }

    essential: list[str] = []
    for m in minterms:
        if len(chart[m]) == 1 and chart[m][0] not in essential:
            essential.append(chart[m][0])

    remaining_primes = [str(prime["pattern"]) for prime in primes]

    def covers_all(patterns: list[str]) -> bool:
        return all(any(covers(pattern, m) for pattern in patterns) for m in minterms)

    candidates: list[list[str]] = []
    for mask in range(1, 1 << len(remaining_primes)):
        patterns = [
            pattern
            for index, pattern in enumerate(remaining_primes)
            if mask & (1 << index)
        ]
        if covers_all(patterns):
            candidates.append(patterns)
    candidates.sort(key=lambda items: (literal_count(items), len(items), ",".join(items)))

    return {
        "minterms": minterms,
        "stages": stages,
        "primes": primes,
        "chart": chart,
        "essential": essential,
        "selected": candidates[0],
    }


def dnf_term(pattern: str) -> str:
    literals = [
        VARS[index] if bit == "1" else f"¬{VARS[index]}"
        for index, bit in enumerate(pattern)
        if bit != "-"
    ]
    return " ∧ ".join(literals) if literals else "1"


def cnf_clause_from_zero(pattern: str) -> str:
    literals = [
        f"¬{VARS[index]}" if bit == "1" else VARS[index]
        for index, bit in enumerate(pattern)
        if bit != "-"
    ]
    return f"({' ∨ '.join(literals)})" if literals else "(0)"


def karnaugh(rows: list[dict[str, int]]) -> list[list[int]]:
    ab_order = ["00", "01", "11", "10"]
    cd_order = ["00", "01", "11", "10"]
    result = []
    for ab in ab_order:
        row = []
        for cd in cd_order:
            a, b = int(ab[0]), int(ab[1])
            c, d = int(cd[0]), int(cd[1])
            row.append(
                next(
                    item["f"]
                    for item in rows
                    if item["A"] == a and item["B"] == b and item["C"] == c and item["D"] == d
                )
            )
        result.append(row)
    return result


def main() -> None:
    task7_rows = rows_for(
        lambda a, b, c, d: {
            "implication": impl(a, b),
            "bxord": xor(b, d),
            "right": logical_and(c, xor(b, d)),
            "f": equiv(impl(a, b), logical_and(c, xor(b, d))),
        }
    )
    task8_rows = rows_for(
        lambda a, b, c, d: {
            "implication": impl(a, b),
            "sheffer": nand(c, b),
            "equivalence": equiv(impl(a, b), nand(c, b)),
            "f": xor(equiv(impl(a, b), nand(c, b)), d),
        }
    )

    task8_minterms = [row["i"] for row in task8_rows if row["f"] == 1]
    task8_zeros = [row["i"] for row in task8_rows if row["f"] == 0]
    task8_dnf = quine(task8_minterms)
    task8_cnf = quine(task8_zeros)
    task9_minterms = [1, 2, 3, 5, 10, 14, 15]
    task9 = quine(task9_minterms)

    results = {
        "task7": {
            "variant": 14,
            "expression": "(A → B) ~ (C ∧ (B ⊕ D))",
            "rows": task7_rows,
            "vector": [row["f"] for row in task7_rows],
        },
        "task8": {
            "variant": 1,
            "expression": "((A → B) ~ (C | B)) ⊕ D",
            "rows": task8_rows,
            "vector": [row["f"] for row in task8_rows],
            "minterms": task8_minterms,
            "zeros": task8_zeros,
            "karnaugh": karnaugh(task8_rows),
            "minDnfPatterns": task8_dnf["selected"],
            "minDnf": " ∨ ".join(dnf_term(pattern) for pattern in task8_dnf["selected"]),
            "minCnfPatterns": task8_cnf["selected"],
            "minCnf": " ∧ ".join(cnf_clause_from_zero(pattern) for pattern in task8_cnf["selected"]),
        },
        "task9": {
            "variant": 2,
            "minterms": task9_minterms,
            "stages": task9["stages"],
            "primeImplicants": task9["primes"],
            "chart": task9["chart"],
            "essential": task9["essential"],
            "selected": task9["selected"],
            "minDnf": " ∨ ".join(dnf_term(pattern) for pattern in task9["selected"]),
        },
    }

    text = "\n".join(
        [
            "Практична робота 3: контрольні результати",
            "",
            f"Задача 7: вектор f = [{', '.join(map(str, results['task7']['vector']))}]",
            "",
            f"Задача 8: мінтерми = ({', '.join(map(str, task8_minterms))}), нулі = ({', '.join(map(str, task8_zeros))})",
            f"  мінДНФ = {results['task8']['minDnf']}",
            f"  мінКНФ = {results['task8']['minCnf']}",
            "",
            f"Задача 9: F = ({', '.join(map(str, task9_minterms))})",
            f"  істотні імпліканти = {', '.join(task9['essential'])}",
            f"  мінДНФ = {results['task9']['minDnf']}",
            "",
        ]
    )

    (OUT_DIR / "pr3_tasks7_9_results.json").write_text(
        json.dumps(results, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )
    (OUT_DIR / "pr3_tasks7_9_results.txt").write_text(text, encoding="utf-8")
    print(text)


if __name__ == "__main__":
    main()
